ledger_device_sdk/src/nbgl.rs

```diff
@@ -0,0 +1,267 @@
+use crate::io::{ApduHeader, Comm, Event, Reply};
+use const_zero::const_zero;
+use core::mem::transmute;
+use ledger_secure_sdk_sys::nbgl_icon_details_t;
+use ledger_secure_sdk_sys::*;
+
+pub struct NbglHome;
+pub struct NbglReview;
+pub struct NbglAddressConfirm;
+
+#[derive(PartialEq)]
+pub enum ReviewStatus {
+    Pending,
+    Validate,
+    Reject,
+}
+
+pub struct Field<'a> {
+    pub name: &'a str,
+    pub value: &'a str,
+}
+
+struct NbglContext {
+    icon: Option<&'static [u8]>,
+    name: [u8; 100],
+    info_contents: [[u8; 20]; 2],
+    review_confirm_string: [u8; 40],
+    review_reject_string: [u8; 40],
+    review_pairs: [ledger_secure_sdk_sys::nbgl_layoutTagValue_t; 10],
+    nb_pairs: u8,
+}
+
+const INFOTYPES: [*const ::core::ffi::c_char; 2] = [
+    "Version\0".as_ptr() as *const ::core::ffi::c_char,
+    "Developer\0".as_ptr() as *const ::core::ffi::c_char,
+];
+
+#[no_mangle]
+pub extern "C" fn recv_and_process_event(return_on_apdu: bool) -> bool {
+    unsafe {
+        if let Some(comm) = COMM_REF.as_mut() {
+            let apdu_received = comm.next_event_ahead::<ApduHeader>();
+            if apdu_received && return_on_apdu {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+#[no_mangle]
+pub static mut G_ux_params: bolos_ux_params_t = unsafe { const_zero!(bolos_ux_params_t) };
+
+static mut CTX: NbglContext = unsafe { const_zero!(NbglContext) };
+static mut COMM_REF: Option<&mut Comm> = None;
+
+impl NbglHome {
+    pub fn new(comm: &mut Comm) -> NbglHome {
+        unsafe {
+            COMM_REF = Some(transmute(comm));
+        }
+        NbglHome {}
+    }
+
+    pub fn app_name(self, app_name: &'static str) -> NbglHome {
+        unsafe {
+            CTX.name[..app_name.len()].copy_from_slice(app_name.as_bytes());
+        }
+        self
+    }
+
+    pub fn icon(self, icon: &'static [u8]) -> NbglHome {
+        unsafe {
+            CTX.icon = Some(icon);
+        }
+        self
+    }
+
+    pub fn info_contents(self, version: &str, author: &str) -> NbglHome {
+        unsafe {
+            CTX.info_contents[0][..version.len()].copy_from_slice(version.as_bytes());
+            CTX.info_contents[1][..author.len()].copy_from_slice(author.as_bytes());
+        }
+        self
+    }
+
+    pub fn show<T: TryFrom<ApduHeader>>(&mut self) -> Event<T>
+    where
+        Reply: From<<T as TryFrom<ApduHeader>>::Error>,
+    {
+        unsafe {
+            loop {
+                match ledger_secure_sdk_sys::sync_nbgl_useCaseHomeAndSettings() {
```

Comment:
A "real" `HomeAndSettings` sync API should be used here, this version was just for proof of concept purposes and is pretty bare bones.